src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp

```diff
@@ -198,9 +198,13 @@ void QMCCostFunctionBatched::getConfigurations(const std::string& aroot)
 //                             if the number of samples is not a multiple of the batch size
 void compute_batch_parameters(int sample_size, int batch_size, int& num_batches, int& final_batch_size)
 {
-  num_batches      = sample_size / batch_size;
+  if (batch_size == 0)
```

Comment:
Number of threads is 8 and walkers per rank is 5
Not an efficient setting, but the code shouldn't crash.

```
WARNING Walkers per rank (5) not divisible by number of crowds (8). This will result in a loss of efficiency.
VMCBatched Driver running with
             total_walkers     = 5
             walkers_per_rank  = [5]
             num_crowds        = 8
  on rank 0, walkers_per_crowd = [1(x5), 0(x3)]
```

Reply:
https://github.com/QMCPACK/qmcpack/blob/8658ea9bec1271069c5779e79180fba0adc3a68e/src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp#L274

The call to compute_batch_parameters uses walkers_per_crowd as the batch size input.

Reply:
I feel better to require calling `compute_batch_parameters` with non-zero `batch_size`. With zero `batch_size`, I consider undefined behavior. We need to add protection at upper level.
1. Need to exclude empty crowds in https://github.com/QMCPACK/qmcpack/blob/8658ea9bec1271069c5779e79180fba0adc3a68e/src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp#L246
2. Need to be smarter about uneven crowd_size. https://github.com/QMCPACK/qmcpack/blob/8658ea9bec1271069c5779e79180fba0adc3a68e/src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp#L258

Reply:
IMO, this is more a matter of documentation and defining the API. When and when can something be zero size? Provided it is clear I think we can manage with either. @markdewing your choice on the route - add comments + migrate the "defense" upwards or keep as is.

Reply:
I believe Mathematician used thousands of years to educate us don't try to find a solution for divide-by-zero because it cannot be defined. If so, we cannot maintain a self-consistent implementation.

Reply:
`compute_batch_parameters` gets called from two places, so pushing the guard up requires changes in two places. And I'm concerned a fix there is going to be much more invasive.

This fix isn't a solution for divide-by-zero, it's a solution to making the `compute_batch_parameters` function defined for all inputs.

Reply:
I'm pretty sure if we don't protect the caller side, more issues will occur down the road and it is challenging to maintain the behavior of 0 of the whole `evalOptConfig` and `evalOptCorrelated`.
I won't call it invasive. For the two places, we should just create helper functions and reuse code.

Reply:
This (`compute_batch_parameters`) *is* that helper function.   I think the solution is to make sure the rest of the code handles zero-sized `num_batches` and `final_batch_size` okay (which I think it does - after the fix there are no crashes and the tests pass).  It seems better to make the regular code path robust/correct to zero size rather than make a special code path for it.

Reply:
I mean need helper functions for implementing my point 1 and 2 I mentioned in https://github.com/QMCPACK/qmcpack/pull/4691#discussion_r1284570262.
handling zero is a can of worms. I'm pretty sure is it not worth doing that. We should avoid calling zero. it is like what is the expected behavior of calling malloc with zero, the recommendation is don't call it.
As I explained, making a crowd behaving as no-op is worse than not using that crowd.

Reply:
In this PR, I prefer not allowing calling to compute_batch_parameters with zero batch_size but I can take a stop-gap mentioned in https://github.com/QMCPACK/qmcpack/pull/4691#discussion_r1284615833. Long term, the code needs to be changed into what I mentioned in https://github.com/QMCPACK/qmcpack/pull/4691#discussion_r1284570262.

Reply:
So in QMCDriverNew.h
We do say:
   *  You can have crowds or ranks with no walkers.

And at one point I certainly a great deal of code so that if the number of walkers per crowd went to 0 it wasn't in and of itself a failure mode since it could happen during population control. 

But as I understand it the crowds in this part of the optimization are separate from the VMCBatch/DMCBatch drivers crowds and should be sized based on the number of samples being used.  I think a sane number of samples is >> number of crowds. So while we need to do something if a batch size ends up being 0 isn't that evidence of some nonsense input and we should just throw an exception and be down with it.

Reply:
The problem with throwing an exception here is that it won't happen until after the VMC section has collected the samples.  If this situation is a problem severe enough that it should halt the program, it should be dealt with during input validation.

Reply:
It seems like everything that one needs to know is available before anything gets done. But this is a weakness in the base input class as well. But the lazy evaluation of the input is pretty undesirable element of the legacy design so I don't think keeping it is a good justification. I think its a weakness in the QMCDriverInput that it doesn't check this. We know everything we need to after a successful parse of the input, its just the input parsing here is still partially legacy style i.e. there isn't really an input class.

How severe this input is a matter of opinion, its indicative of a badly setup calculation that is likely wasting collective resources. If we ever do have nested threading I don't think we want to specify it be having number of crowds > number of walkers. 

I think it should abort back in the VMC.  If you don't have the resources to at least run number of walkers == number of crowds you should run fewer crowds.

I think if we just generally disallow this sort of nonsense setup for drivers then you don't have to deal with this issue.